### ISSUE TITLE
Twig Variables/Data in Pimcore 11

### ISSUE THREAD

**Issue by fashxp** [pimcore/pimcore#16474]
### Discussed in https://github.com/pimcore/pimcore/discussions/16407

<div type='discussions-op-text'>

<sup>Originally posted by **limenet** January  2, 2024</sup>
While upgrading a project from Pimcore 10 to 11 I encountered an unexpected behavior when passing variables to the Twig templates.

---

**Pimcore 10**

Up until Pimcore 10, we'd write our controller actions like this:

```php
/**
 * @Template
 */
public function someAction(SomeService $someService): array
{
    return [
        'data' => $someService->getData()
    ];
}
```

And in the Twig template we'd have access to `data`, e.g.

```twig
{{ data }}
```

This worked no matter if we specified the template in the annotation, used the auto-discovery feature (controller + action), or specified/overwrote the template in the Pimcore Document.

---

**Pimcore 11**

As template auto-discovery is no longer supported in Pimcore 11, we switched to attributes:

```php
#[Template('some.html.twig')]
public function someAction(SomeService $someService): array
{
    return [
        'data' => $someService->getData()
    ];
}
```

Now, however, the behavior is different based on whether the template is specified/overwritten on the `Document` instance:

- `$document->getTemplate() === null`
  - Behavior remains unchanged and `{{ data }}` works.
  - Symfony logic in `\Symfony\Bridge\Twig\EventListener\TemplateAttributeListener` is used
- `$document->getTemplate() !== null`
  - We no longer have access to `data` inside our Twig template
  - `someService` is now available in the Twig template
  - Pimcore logic in `\Pimcore\Bundle\CoreBundle\EventListener\Frontend\ContentTemplateListener` is used

---

To restore the old behavior, the following patch can be applied:

```patch
--- a/bundles/CoreBundle/src/EventListener/Frontend/ContentTemplateListener.php
+++ b/bundles/CoreBundle/src/EventListener/Frontend/ContentTemplateListener.php
@@ -58,6 +58,7 @@
     public function onKernelView(ViewEvent $event): void
     {
         $request = $event->getRequest();
+        $parameters = $event->getControllerResult();

         if (!$this->matchesPimcoreContext($request, PimcoreContextResolver::CONTEXT_DEFAULT)) {
             return;
@@ -70,7 +71,7 @@
             return;
         }

-        $parameters = $this->resolveParameters($event->controllerArgumentsEvent, $attribute?->vars);
+        $parameters ??= $this->resolveParameters($event->controllerArgumentsEvent, $attribute?->vars);
         $status = 200;

         if (interface_exists('Symfony\\Component\\Form\\FormInterface')) {
```
---

**Question**: Is this new behavior intentional or am I missing something?</div>

**Comment by fashxp** [comment]
@brusch wdyt?

**Comment by brusch** [comment]
@fashxp LGTM - can't think of any particular reason for that, so I'd assume it's unintentional behavior and therefore a bug.

**Comment by kingjia90** [comment]
Looks like resolved by #16456 (11.1.5) but it's not released yet
https://github.com/pimcore/pimcore/pull/16456/files#diff-2f932cea0ff8fb705f8b326c34e77f71766bf96d6b9a40ac2d31371580d5f1deR96

**Comment by kingjia90** [comment]
Fixed by https://github.com/pimcore/pimcore/pull/16456
